README.md
# abnfu


ChangeLog.md
# Changelog for abnfu

## Unreleased changes


test/Spec.hs
import           Test.Tasty (TestTree)
import qualified Test.Tasty as T

import qualified Test.ABNFU.ABNF.ParserTest (tests)

main :: IO ()
main = T.defaultMain tests

tests :: TestTree
tests = T.testGroup "All tests"
    [ Test.ABNFU.ABNF.ParserTest.tests
    ]


test/Test/ABNFU/ABNF/ParserTest.hs
{-# LANGUAGE OverloadedStrings #-}
module Test.ABNFU.ABNF.ParserTest where

import qualified ABNFU.ABNF.Grammar     as G
import           ABNFU.ABNF.Parser      (abnf, block, comment, literalChars,
                                         literalString, pElem, repeats, rule,
                                         ruleName)

import           Control.Monad.IO.Class (liftIO)
import qualified Data.List.NonEmpty     as NE
import qualified Data.Text              as T
import qualified Data.Text.IO           as T

import           Hedgehog               (Property, property, withTests, (===))
import           Test.Tasty             (TestTree)
import           Test.Tasty             (testGroup)
import           Test.Tasty.Hedgehog    (testProperty)
import           Text.Megaparsec        (parseMaybe)


tests :: TestTree
tests = testGroup "ABNFU.ABNF.Grammar"
    [ testProperty "unit (IO): abnf core"    unit_abnfcore
    , testProperty "unit (IO): abnf in abnf" unit_abnfabnf
    , testProperty "unit: block"             unit_block
    , testProperty "unit: rule"              unit_rule
    , testProperty "unit: literalChars"      unit_literalChars
    , testProperty "unit: ruleName"          unit_ruleName
    , testProperty "unit: pElem"             unit_pElem
    , testProperty "unit: literalString"     unit_literalString
    , testProperty "unit: repeats"           unit_repeats
    , testProperty "unit: comment"           unit_comment
    ]


unit_abnfcore :: Property
unit_abnfcore = withTests 1 $ property $ do
    core <- liftIO $ T.readFile "doc/core.abnf"
    case parseMaybe abnf core of
        Just _  -> pure ()
        Nothing -> fail "Could not parse core ABNF: doc/core.abnf"


unit_abnfabnf :: Property
unit_abnfabnf = withTests 1 $ property $ do
    gabnf <- liftIO $ T.readFile "doc/abnf.abnf"
    case parseMaybe abnf gabnf of
        Just _  -> pure ()
        Nothing -> fail "Could not parse ABNF definition: doc/abnf.abnf"


unit_block :: Property
unit_block = withTests 1 $ property $ do

    let e1 = G.ElemString (G.LiteralString Nothing "hello")
    let e2 = G.ElemOptional e1
    let e3 = G.ElemChars (G.LiteralChars G.Hexadecimal (G.CharsRange 48 57))
    let e4 = G.ElemConcat e2 e3
    let r1 = G.RuleBase (G.RuleName "rule-name") e4

    let b1 = G.BlockRule r1
    let i1 = T.unlines
             [ "rule-name ="
             , "  [\"hello\"]  ; literal string hello"
             , "  %x30-39      ; character range"
             , ""
             ]
    parseMaybe block i1 === Just (Right b1)

    let b2 = G.BlockLineComment (G.Comment " this is a comment")
    parseMaybe block "  ; this is a comment\n" === Just (Right b2)

    parseMaybe block "  \n" === Just (Left ())


unit_rule :: Property
unit_rule = withTests 1 $ property $ do

    let e1 = G.ElemString (G.LiteralString Nothing "hello")
    let e2 = G.ElemOptional e1
    let e3 = G.ElemChars (G.LiteralChars G.Hexadecimal (G.CharsRange 48 57))
    let e4 = G.ElemConcat e2 e3

    let r1 = G.RuleBase (G.RuleName "rule-name") e4
    parseMaybe rule "rule-name   =   [\"hello\"] %x30-39 \n" === Just r1

    let r2 = G.RuleIncremental (G.RuleName "rule-name") e4
    parseMaybe rule "rule-name /= [\"hello\"] %x30-39 ; comment\n" === Just r2


unit_ruleName :: Property
unit_ruleName = withTests 1 $ property $ do
    parseMaybe ruleName "foo-bar42" === Just (G.RuleName "foo-bar42")


unit_pElem :: Property
unit_pElem = withTests 1 $ property $ do

    let e1 = G.ElemNamedRule (G.RuleName "rule-name")
    parseMaybe pElem "rule-name" === Just e1

    let c1 = G.LiteralChars G.Hexadecimal (G.CharsRange 48 57)
    let e2 = G.ElemChars c1
    parseMaybe pElem "%x30-39  " === Just e2

    let e3 = G.ElemString (G.LiteralString Nothing "hello")
    parseMaybe pElem "\"hello\" " === Just e3

    let e4 = G.ElemRepeat (G.RepeatsBetween 3 5) e1
    parseMaybe pElem "3*5rule-name " === Just e4

    let e5 = G.ElemOptional e3
    parseMaybe pElem "[ \"hello\" ] " === Just e5

    let e6 = G.ElemConcat e5 e1
    parseMaybe pElem "[\"hello\"] \n rule-name" === Just e6
    parseMaybe pElem "[\"hello\"]  ; comment \n rule-name" === Just e6

    let e7 = G.ElemAlternative e6 e4
    parseMaybe pElem "[\"hello\"] rule-name / 3*5rule-name" === Just e7

    let e8 = G.ElemAlternative (G.ElemParen e6) e4
    parseMaybe pElem "( [\"hello\"] rule-name ) / 3*5rule-name" === Just e8

    let e9 = G.ElemConcat e3 e2
    parseMaybe pElem "\"hello\"\n  %x30-39" === Just e9


unit_literalChars :: Property
unit_literalChars = withTests 1 $ property $ do

    -- RFC-5234 2.3
    let d13 = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [13]))
    let x0d = G.LiteralChars G.Hexadecimal (G.CharsList (NE.fromList [13]))
    let dcrlf = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [13, 10]))
    let d97 = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [97]))
    let d98 = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [98]))
    let d99 = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [99]))
    let d9799 = G.LiteralChars G.Decimal (G.CharsList (NE.fromList [97, 98, 99]))
    parseMaybe literalChars "%d13" === Just d13
    parseMaybe literalChars "%x0D" === Just x0d
    parseMaybe literalChars "%d13.10" === Just dcrlf
    parseMaybe literalChars "%d97" === Just d97
    parseMaybe literalChars "%d98" === Just d98
    parseMaybe literalChars "%d99" === Just d99
    parseMaybe literalChars "%d97.98.99" === Just d9799

    -- RFC-5234 3.1
    let foo = G.LiteralChars G.Hexadecimal (G.CharsList (NE.fromList [97]))
    let bar = G.LiteralChars G.Hexadecimal (G.CharsList (NE.fromList [98]))
    parseMaybe literalChars "%x61" === Just foo
    parseMaybe literalChars "%x62" === Just bar

    -- RFC-5234 3.4
    let digit = G.LiteralChars G.Hexadecimal (G.CharsRange 48 57)
    parseMaybe literalChars "%x30-39" === Just digit

    -- RFC-5234 3.4
    let x0d0a = G.LiteralChars G.Hexadecimal (G.CharsList (NE.fromList [13, 10]))
    let x207e = G.LiteralChars G.Hexadecimal (G.CharsRange 32 126)
    parseMaybe literalChars "%x0D.0A" === Just x0d0a
    parseMaybe literalChars "%x20-7E" === Just x207e


unit_literalString :: Property
unit_literalString = withTests 1 $ property $ do

    let hello = G.LiteralString Nothing "hello"
    parseMaybe literalString "\"hello\"" === Just hello

    let empty = G.LiteralString Nothing ""
    parseMaybe literalString "\"\"" === Just empty

    let ciHello = G.LiteralString (Just G.CaseInsensitive) "hello"
    parseMaybe literalString "%i\"hello\"" === Just ciHello

    let csHello = G.LiteralString (Just G.CaseSensitive) "hello"
    parseMaybe literalString "%s\"hello\"" === Just csHello


unit_repeats :: Property
unit_repeats = withTests 1 $ property $ do
    parseMaybe repeats "5"   === Just (G.RepeatsExactly 5)
    parseMaybe repeats "3*"  === Just (G.RepeatsAtLeast 3)
    parseMaybe repeats "*5"  === Just (G.RepeatsAtMost 5)
    parseMaybe repeats "3*5" === Just (G.RepeatsBetween 3 5)
    parseMaybe repeats "*"   === Just (G.RepeatsAny)


unit_comment :: Property
unit_comment = withTests 1 $ property $ do
    parseMaybe comment ";\n" === Just (G.Comment "")
    parseMaybe comment "; foo bar\t  \n" === Just (G.Comment " foo bar\t  ")


test/Test/ABNFU/ABNF/GrammarTest.hs
module Test.ABNFU.ABNF.GrammarTest where

import Hedgehog (Property, property, withTests)
import Test.Tasty (TestTree)
import Test.Tasty.Hedgehog (testProperty)
import Test.Tasty (testGroup)

tests :: TestTree
tests = testGroup "ABNFU.ABNF.Grammar"
    [ testProperty "unit: literalChars" unit_literalChars
    ]

unit_literalChars :: Property
unit_literalChars = withTests 1 $ property $ do
    pure ()


src/ABNFU/ABNF.hs
{-|
Module      : ABNFU.ABNF
Description : ABNF grammar and parser.

Data types for the components of an ABNF grammar ("as written"), and a parser
for ABNF text files, corresponding to RFC-5234 and RFC-7405. All parts of ABNF
are supported except free-form prose.
-}
module ABNFU.ABNF
    ( -- * ABNF AST types
      G.ABNFGrammar(..)
    , G.Block(..)
    , G.Comment(..)
    , G.Rule(..)
    , G.Elem(..)
    , G.RuleName(..)
    , G.LiteralChars(..)
    , G.Chars(..)
    , G.Base(..)
    , G.LiteralString(..)
    , G.CaseSensitivity(..)
    , G.Repeats(..)
      -- * Parser for ABNF
    , ABNFParser
    , parseABNF
    ) where

import qualified ABNFU.ABNF.Grammar as G
import qualified ABNFU.ABNF.Parser as P

-- | Parser type for ABNF.
type ABNFParser = P.Parser

-- | Parses an ABNF file producing a sequence of blocks.
parseABNF :: P.Parser [G.Block]
parseABNF = P.abnf


src/ABNFU/ABNF/Grammar.hs
{-|
Module      : ABNFU.ABNF.Grammar
Description : Haskell data types to describe an ABNF grammar.

The types in this module describe the components of an ABNF grammar
"as written", corresponding to RFC-5234 and RFC-7405. All parts of ABNF are
represented except free-form prose.
-}
{-# OPTIONS_HADDOCK hide #-}
module ABNFU.ABNF.Grammar
    ( -- * Types
      ABNFGrammar(..)
    , Block(..)
    , Comment(..)
    , Rule(..)
    , Elem(..)
    , RuleName(..)
    , LiteralChars(..)
    , Chars(..)
    , Base(..)
    , LiteralString(..)
    , CaseSensitivity(..)
    , Repeats(..)
    ) where

import           Data.CaseInsensitive (CI)
import           Data.List.NonEmpty   (NonEmpty)
import           Data.Text            (Text)


-- | Parsed contents of an ABNF grammar.
newtype ABNFGrammar = ABNFGrammar [Block]


-- | ABNF block.
data Block
    -- | Isolated comment on a line by itself.
    --
    --   Comments that appear elsewhere are discarded for now.
    = BlockLineComment !Comment
    -- | Rule.
    | BlockRule !Rule
    deriving (Eq, Show)


newtype Comment
    = Comment Text
    deriving (Eq, Show)


-- | ABNF rule.
data Rule
    -- | Base rule (RFC-5234 2.2).
    --
    --   > name = elements crlf
    = RuleBase !RuleName !Elem
    -- | Incremental alternatives for a rule (RFC-5234 3.3).
    --
    --   > oldrule =/ additional-alternatives
    | RuleIncremental !RuleName !Elem
    deriving (Eq, Show)


-- | Expression language of rule elements.
data Elem
    -- | Uses a named rule.
    --
    --   > name1 = name2
    = ElemNamedRule !RuleName
    -- | A character literal (RFC-5234 2.3).
    --
    --   > %x61     ; 'a' (single character)
    --   > %d13.10  ; multiple characters, separated by periods
    | ElemChars !LiteralChars
    -- | A string literal (RFC-5234 2.3 and RFC-7405).
    --
    --   > "command string"    ; case-insensitive string (RFC-5234)
    --   > %i"command string"  ; case-insensitive string (RFC-7405)
    --   > %s"command string"  ; case-sensitive string (RFC-7405)
    | ElemString !LiteralString
    -- | Concatenation of elements (RFC-5234 3.1).
    --
    --   > rule = a b c
    | ElemConcat !Elem !Elem
    -- | Alternative elements (RFC-5234 3.2).
    --
    --   > rule = a / b
    | ElemAlternative !Elem !Elem
    -- | Repeated element (RFC-5234 3.6, 3.7 and 3.8).
    --
    --   > 5e     ; repeats element e 5 times
    --   > 2*5e   ; repeats e a minimum of 2 and maximum of 5 times
    --   > *5e    ; repeats e from zero to 5 times
    --   > [e]    ; e is optional
    | ElemRepeat Repeats !Elem
    -- | Parenthesized elements (RFC-5234 3.5).
    --
    --   > (a b)
    | ElemParen !Elem
    -- | Optional elements (RFC-5234 3.8).
    --
    --   > [foo bar]
    | ElemOptional !Elem
    deriving (Eq, Show)


-- | Name of a rule (RFC-5234 2.1).
newtype RuleName
    = RuleName (CI Text)
    deriving (Eq, Show)


-- | Numeric representation of terminal character(s) (RFC-5234 2.3, 3.4).
data LiteralChars
    = LiteralChars !Base !Chars
    deriving (Eq, Show)


-- | Specification of characters.
data Chars
    -- | Period-separated list of single character (RFC-5234 2.3).
    --
    --   These are concatenated characters.
    = CharsList !(NonEmpty Integer)
    -- | Inclusive range of characters (RFC-5234 3.4).
    | CharsRange !Integer !Integer
    deriving (Eq, Show)


-- | Base in which a numeric character is expressed (RFC-5234 2.3).
data Base
    = Binary
    | Decimal
    | Hexadecimal
    deriving (Eq, Show)


-- | Literal string (RFC-5234 2.3).
data LiteralString
    = LiteralString !(Maybe CaseSensitivity) !Text
    deriving (Eq, Show)


-- | Case sensitivity of a literal string (RFC-7405).
data CaseSensitivity
    = CaseSensitive
    | CaseInsensitive
    deriving (Eq, Show)


-- | Specifies how repeats of an element occur (RFC-5234 3.6, 3.7 and 3.8).
--
--   This covers @<a>*<b>element@ notation.
data Repeats
    -- | Specifies exactly how many times an element should repeat.
    --
    --   > 5e
    = RepeatsExactly !Integer
    -- | Specifies the minimum number of times an element should repeat.
    --
    --   > 3*e
    | RepeatsAtLeast !Integer
    -- | Specifies the maximum number of times an element should repeat.
    --
    --   > *5e
    | RepeatsAtMost !Integer
    -- | Specifies a range of times an element should repeat.
    --
    --   > 3*5e
    | RepeatsBetween !Integer !Integer
    -- | Specifies any number of repeats.
    --
    --   > *e
    | RepeatsAny
    deriving (Eq, Show)


src/ABNFU/ABNF/Parser.hs
{-|
Module      : ABNFU.ABNF.Parser
Description : Megaparsec parser for ABNF grammars.

Parses ABNF grammars according to the rules of:
  - RFC-5234: Augmented BNF for Syntax Specifications: ABNF
  - RFC-7405: Case-Sensitive String Support in ABNF

All parts of ABNF except free-form prose are understood.
-}
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_HADDOCK hide #-}
module ABNFU.ABNF.Parser
    ( -- * Parser type
      Parser
      -- * Functions
    , abnf
    , block
    , rule
    , ruleName
    , pElem
    , literalChars
    , literalString
    , repeats
    , comment
    , cwsp
    , cnl
    ) where

import qualified ABNFU.ABNF.Grammar        as G

import           Control.Monad.Combinators (many, optional, sepBy1, some)
import qualified Data.CaseInsensitive      as CI
import           Data.Char                 (isAsciiLower, isDigit, isHexDigit,
                                            ord)
import           Data.Either               (rights)
import qualified Data.List.NonEmpty        as NE
import           Data.Text                 (Text)
import qualified Data.Text                 as T
import           Data.Void                 (Void)
import           Numeric                   (readDec, readHex, readInt)
import           Text.Megaparsec           (Parsec, takeWhile1P, takeWhileP,
                                            try, (<|>))
import           Text.Megaparsec.Char      (char, char', satisfy, string,
                                            string')


type Parser = Parsec Void Text


-- | Parses a sequence of blocks forming an ABNF grammar.
abnf :: Parser [G.Block]
abnf = rights <$> many block


block :: Parser (Either () G.Block)
block =
        (Left <$> try (many wsp *> nlChars))
    <|> ((Right . G.BlockLineComment) <$> try (many wsp *> comment))
    <|> ((Right . G.BlockRule) <$> rule)


rule :: Parser G.Rule
rule = try ruleBase <|> try ruleIncremental

  where

    ruleBase :: Parser G.Rule
    ruleBase
        =   G.RuleBase
        <$> (ruleName <* many cwsp <* char '=' <* many cwsp)
        <*> pElem <* cnl

    ruleIncremental :: Parser G.Rule
    ruleIncremental
        =   G.RuleIncremental
        <$> (ruleName <* many cwsp <* string "/=" <* many cwsp)
        <*> pElem <* cnl



ruleName :: Parser G.RuleName
ruleName =
    (G.RuleName . CI.mk)
    <$> (T.cons <$> satisfy isAsciiAlpha <*> takeWhileP Nothing isNameChar)

  where

    isAsciiAlpha :: Char -> Bool
    isAsciiAlpha c =
        let
            c' = ord c
        in
               (c' >= 65 && c' <= 90)
            || (c' >= 97 && c' <= 122)

    isNameChar :: Char -> Bool
    isNameChar c = isAsciiAlpha c || isDigit c || c == '-'


pElem :: Parser G.Elem
pElem = alternation <* many wsp  {- note: many wsp is from the erratum -}

  where

    alternation :: Parser G.Elem
    alternation =
            try (    G.ElemAlternative
                 <$> (concatenation <* many cwsp <* char '/' <* many cwsp)
                 <*> alternation
                )
        <|> concatenation

    concatenation :: Parser G.Elem
    concatenation =
            try (G.ElemConcat <$> (repetition <* some cwsp) <*> concatenation)
        <|> repetition

    repetition :: Parser G.Elem
    repetition =
            try (G.ElemRepeat <$> repeats <*> element)
        <|> element

    element :: Parser G.Elem
    element =
            terminal
        <|> group
        <|> option

    group :: Parser G.Elem
    group = bracketed '(' ')' (G.ElemParen <$> alternation)

    option :: Parser G.Elem
    option = bracketed '[' ']' (G.ElemOptional <$> alternation)

    terminal :: Parser G.Elem
    terminal =
            (G.ElemChars <$> literalChars)
        <|> (G.ElemString <$> literalString)
        <|> (G.ElemNamedRule <$> ruleName)

    bracketed :: Char -> Char -> Parser a -> Parser a
    bracketed c1 c2 p =
        char c1 *> many cwsp *> p <* many cwsp <* char c2


-- | Whitespace, or a newline and then more whitespace.
cwsp :: Parser ()
cwsp =
    (wsp <|> (cnl *> wsp))
    *> pure ()


-- | Comment or one or more characters of a newline.
cnl :: Parser ()
cnl = (comment *> pure ()) <|> nlChars


-- | One or more characters of a newline.
nlChars :: Parser ()
nlChars = takeWhile1P Nothing isNL *> pure ()
  where
    isNL c = (c == '\n') || (c == '\r')


-- | Single character of whitespace.
wsp :: Parser ()
wsp = (char ' ' <|> char '\t') *> pure ()


literalChars :: Parser G.LiteralChars
literalChars = do
    _ <- char '%'
    b <- base
    cs <- case b of
        G.Binary      -> chars binVal
        G.Decimal     -> chars integer
        G.Hexadecimal -> chars hexVal
    pure (G.LiteralChars b cs)

  where

    base :: Parser G.Base
    base =
            (char' 'b' *> pure G.Binary)
        <|> (char' 'd' *> pure G.Decimal)
        <|> (char' 'x' *> pure G.Hexadecimal)

    chars :: Parser Integer -> Parser G.Chars
    chars p =
            try (G.CharsRange <$> (p <* char '-') <*> p)
        <|> ((G.CharsList . NE.fromList) <$> p `sepBy1` char '.')

    binVal :: Parser Integer
    binVal = do
        let isBinDigit c = c == '0' || c == '1'
        let toInt c = case c of
                '0' -> 0
                '1' -> 1
                _   -> error "unexpected binary character"
        txt <- takeWhile1P Nothing isBinDigit
        runUniqueReadS (readInt 2 (const True) toInt) txt

    hexVal :: Parser Integer
    hexVal = takeWhile1P Nothing isUpperCaseHex >>= runUniqueReadS readHex

    isUpperCaseHex :: Char -> Bool
    isUpperCaseHex c = isHexDigit c && (not $ isAsciiLower c)


literalString :: Parser G.LiteralString
literalString = G.LiteralString <$> caseSensitivity <*> quotedString

  where

    caseSensitivity :: Parser (Maybe G.CaseSensitivity)
    caseSensitivity =
        optional
        (    (string' "%i" *> pure G.CaseInsensitive)
         <|> (string' "%s" *> pure G.CaseSensitive)
        )

    quotedString :: Parser Text
    quotedString = char '"' *> takeWhileP Nothing isStringContent <* char '"'

    isStringContent :: Char -> Bool
    isStringContent c =
        let
            c' = ord c
        in
            (c' == 32) || (c' == 33) || (c' >= 35 && c' <= 126)


repeats :: Parser G.Repeats
repeats =
        try (G.RepeatsBetween <$> (integer <* char '*') <*> integer)
    <|> try (G.RepeatsAtLeast <$> (integer <* char '*'))
    <|> try (G.RepeatsAtMost  <$> (char '*' *> integer))
    <|> try (G.RepeatsExactly <$> integer)
    <|> try (char '*' *> pure G.RepeatsAny)


comment :: Parser G.Comment
comment = char ';' *> (G.Comment <$> takeWhileP Nothing isVCharOrWS) <* nlChars
  where
    isVCharOrWS :: Char -> Bool
    isVCharOrWS c =
        let
            c' = ord c
        in
            (c == ' ') || (c == '\t') || (c' >= 33 && c' <= 126)


-- | Parses a decimal integer.
integer :: Parser Integer
integer = takeWhile1P Nothing isDigit >>= runUniqueReadS readDec


-- | Runs a 'ReadS' parser on a given piece of 'Text', and fails if there is
--   not a unique parse result.
runUniqueReadS :: ReadS a -> Text -> Parser a
runUniqueReadS rs txt =
    case (rs . T.unpack) txt of
        [ (x, "") ] -> pure x
        _           -> fail "unexpected parse error in runUniqueReadS"
